src/flow/abstractObjectFactories.js

```diff
@@ -0,0 +1,124 @@
+import { ObjectCreate, ArrayCreate } from "../methods/index.js";
+import { ValuesDomain } from "../domains/index.js";
+import { Value, AbstractValue, ObjectValue, ArrayValue } from "../values/index.js";
+import invariant from "../invariant.js";
+
+type ObjectTypes = {
+  [key: string]: ObjectTypes | string,
+};
+
+export function createObject(realm: Realm, shape: ObjectTypes | null, name: string | null): ObjectValue {
+  let obj = ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+  if (shape != null) {
+    // to get around Flow complaining that shape could be null
+    let shapeThatIsNotNull = shape;
+    Object.keys(shape).forEach((id: string) => {
+      let value = shapeThatIsNotNull[id];
+      invariant(value instanceof Value, "creation of object failed due to object containing non-value properties");
+      obj.$Set(id, value, obj);
+      if (name !== null) {
+        value.intrinsicName = `${name}.${id}`;
+      }
+    });
+  }
+  if (name !== null) {
+    obj.intrinsicName = name;
+  }
+  return obj;
+}
+
+export function createArray(realm: Realm, name: string | null): ArrayValue {
+  let obj = ArrayCreate(realm, 0, realm.intrinsics.ArrayPrototype);
+  if (name !== null) {
+    obj.intrinsicName = name;
+  }
+  return ((obj: any): ArrayValue);
+}
+
+function _createAbstractArray(realm: Realm, name: string | null): AbstractValue {
+  if (name === null) {
+    name = "unknown";
+  }
+  let value = AbstractValue.createFromTemplate(realm, buildExpressionTemplate(name), ArrayValue, [], name);
+  value.intrinsicName = name;
+  let template = createArray(realm, name);
+  template.makePartial();
+  template.makeSimple();
+  value.values = new ValuesDomain(new Set([template]));
+  realm.rebuildNestedProperties(value, name);
+  return value;
+}
+
+function _createAbstractObject(realm: Realm, name: string | null, objectTypes: ObjectTypes | null): AbstractValue {
+  if (name === null) {
+    name = "unknown";
+  }
+  let value = AbstractValue.createFromTemplate(realm, buildExpressionTemplate(name), ObjectValue, [], name);
+  value.intrinsicName = name;
+  let template = createObject(realm, objectTypes, name);
+  template.makePartial();
+  template.makeSimple();
+  value.values = new ValuesDomain(new Set([template]));
+  realm.rebuildNestedProperties(value, name);
+  return value;
+}
+
+export function createAbstractObject(
+  realm: Realm,
+  name: string | null,
+  objectTypes: ObjectTypes | null | string
+): ObjectValue | AbstractValue {
+  if (typeof objectTypes === "string") {
+    invariant(
+      objectTypes === "empty" || objectTypes === "object",
+      `Expected an object or a string of "empty" or "object" for createAbstractObject() paramater "objectTypes"`
+    );
+    return _createAbstractObject(realm, name, null);
+  }
+  if (objectTypes !== null) {
+    let propTypeObject = {};
+
+    Object.keys(objectTypes).forEach(key => {
+      let value = ((objectTypes: any): ObjectTypes)[key];
```

Comment:
Prefer 
```javascript
let objTypes = objectTypes;
invariant(objTypes);
```
or something similar to any cast if possible. (Flow is bad at tracking nullability through captures and usually only refines local variables.)